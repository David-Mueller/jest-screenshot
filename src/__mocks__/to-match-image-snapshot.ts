export const toMatchImageSnapshot = jest.fn();
