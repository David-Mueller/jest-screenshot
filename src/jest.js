"use strict";
exports.__esModule = true;
function isSnapshotState(obj) {
    if (typeof obj !== "object") {
        return false;
    }
    var _counters = obj._counters;
    if (typeof _counters !== "object") {
        return false;
    }
    if (typeof _counters.get !== "function") {
        return false;
    }
    return true;
}
exports.isSnapshotState = isSnapshotState;
function isJestTestConfiguration(obj) {
    if (typeof obj !== "object") {
        return false;
    }
    var snapshotState = obj.snapshotState, testPath = obj.testPath, currentTestName = obj.currentTestName, isNot = obj.isNot;
    if (typeof testPath !== "string") {
        return false;
    }
    if (typeof currentTestName !== "string") {
        return false;
    }
    if (typeof isNot !== "boolean") {
        return false;
    }
    if (!isSnapshotState(snapshotState)) {
        return false;
    }
    return true;
}
exports.isJestTestConfiguration = isJestTestConfiguration;
