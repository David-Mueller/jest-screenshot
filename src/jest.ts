export interface SnapshotState {
    _counters: Map<string, number>;
    _updateSnapshot: "new" | "all" | "none";
}

export interface JestTestConfiguration {
    snapshotState: SnapshotState;
    testPath: string;
    currentTestName: string;
    isNot: boolean;
}

export interface MatcherResult {
    message?(): string;
    pass: boolean;
}

export function isSnapshotState(obj: any): obj is SnapshotState {
    if (typeof obj !== "object") { return false; }
    const { _counters } = obj;
    if (typeof _counters !== "object") { return false; }
    if (typeof _counters.get !== "function") { return false; }
    return true;
}

export function isJestTestConfiguration(obj: any): obj is JestTestConfiguration {
    if (typeof obj !== "object") { return false; }
    const { snapshotState, testPath, currentTestName, isNot } = obj;
    if (typeof testPath !== "string") { return false; }
    if (typeof currentTestName !== "string") { return false; }
    if (typeof isNot !== "boolean") { return false; }
    if (!isSnapshotState(snapshotState)) { return false; }
    return true;
}
