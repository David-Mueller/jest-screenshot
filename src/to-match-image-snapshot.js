"use strict";
exports.__esModule = true;
var node_libpng_1 = require("node-libpng");
var native_image_diff_1 = require("native-image-diff");
var jest_1 = require("./jest");
var path = require("path");
var lodash_kebabcase_1 = require("lodash.kebabcase");
function getSnapshotFileName(testPath, currentTestName, snapshotState, _a) {
    var identifier = _a.identifier;
    var counter = (snapshotState._counters.get(currentTestName) || 0) + 1;
    if (typeof identifier === "function") {
        var fileName = identifier(testPath, currentTestName, counter);
        if (fileName.toLowerCase().endsWith(".snap.png")) {
            return fileName;
        }
        return fileName + ".snap.png";
    }
    if (typeof identifier !== "undefined") {
        throw new Error("Jest: Invalid configuration for `.toMatchImageSnapshot`: `identifier` must be a function.");
    }
    return lodash_kebabcase_1["default"](path.basename(testPath)) + "-" + lodash_kebabcase_1["default"](currentTestName) + "-" + counter;
}
function getSnapshotPath(testPath, currentTestName, snapshotState, configuration) {
    var snapshotsDir = configuration.snapshotsDir;
    var fileName = getSnapshotFileName(testPath, currentTestName, snapshotState, configuration);
    return path.join(path.dirname(testPath), snapshotsDir || "__snapshots__", fileName);
}
function toMatchImageSnapshot(received, configuration) {
    if (!jest_1.isJestTestConfiguration(this)) {
        throw new Error("Jest: Attempted to call toMatchImageSnapshot outside of Jest context.");
    }
    var _a = this, testPath = _a.testPath, currentTestName = _a.currentTestName, isNot = _a.isNot;
    var snapshotState = this.snapshotState;
    console.log(snapshotState, this);
    if (isNot) {
        throw new Error("Jest: `.not` cannot be used with `.toMatchImageSnapshot()`.");
    }
    var colorThreshold = configuration.colorThreshold, detectAntialiasing = configuration.detectAntialiasing;
    native_image_diff_1.diffImages({
        image1: node_libpng_1.decode(received),
        image2: node_libpng_1.readPngFileSync(getSnapshotPath(testPath, currentTestName, snapshotState, configuration)),
        colorThreshold: colorThreshold,
        detectAntialiasing: detectAntialiasing
    });
}
exports.toMatchImageSnapshot = toMatchImageSnapshot;
function configureToMatchImageSnapshot(configuration) {
    if (configuration === void 0) { configuration = {}; }
    return function (received) {
        return toMatchImageSnapshot(received, configuration);
    };
}
exports.configureToMatchImageSnapshot = configureToMatchImageSnapshot;
