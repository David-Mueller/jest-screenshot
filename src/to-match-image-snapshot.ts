import { decode, readPngFileSync, writePngFileSync, PngImage } from "node-libpng";
import { diffImages } from "native-image-diff";
import chalk from "chalk";
import { existsSync, writeFileSync, readFileSync } from "fs";
import { getSnapshotPath, getReportPath } from "./filenames";
import { SnapshotState, isJestTestConfiguration, MatcherResult } from "./jest";
import { sync as mkdirp } from "mkdirp";
import { dirname, join } from "path";

/**
 * Used as configuration for `toMatchImageSnapshot`.
 */
export interface ToMatchImageSnapshotConfiguration {
    /**
     * Passed to **native-image-diff**. Will disable or enable antialiasing detection.
     * Defaults to `true`.
     */
    detectAntialiasing?: boolean;
    /**
     * Passed to **native-image-diff**. Specifies the threshold on a scale from `0` to `1`
     * for when a pixel counts as changed. `0` allows no difference between two pixels and
     * `1` detects no difference between a white and a black pixel.
     */
    colorThreshold?: number;
    /**
     * If specified, makes the test check for the absolute number of pixels changed. When for example
     * set to `100`, An image which differs from it's snapshot by `101` pixels would fail.
     *
     * Is set to `undefined` by default and hence the check is disabled.
     */
    pixelThresholdAbsolute?: number;
    /**
     * If specified, makes the test check for the relative number of pixels changed. When for example
     * set to `0.5`, An image which differs from it's snapshot by 50.0001% of the pixels would fail.
     *
     * Is set to `0` if neither `pixelThresholdAbsolute` nor `pixelThresholdRelative` are confiured.
     */
    pixelThresholdRelative?: number;
    /**
     * An optional generator function for generating the names for the image snapshot files.
     */
    identifier?: (testPath: string, currentTestName: string, counter: number) => string;
    /**
     * An optional directory name to store the snapshots in. Defaults to `__snapshots__`.
     */
    snapshotsDir?: string;
    /**
     * The directory to write the report to. Defaults to a directory `jest-screenshot-reports` in
     * the projects root directory. Can be set to `null` to explicitly disable generating reports.
     */
    reportPath?: (testPath: string, currentTestName: string, counter: number) => string;
}

/**
 * Performs the actual check for equality of two images.
 *
 * @param snapshotImage The image from the snapshot.
 * @param receivedImage The image received from the `expect(...)` call.
 * @param snapshotNumber The number of the snapshot in this test.
 * @param configuration The configuration of the call to `toMatchImageSnapshot`.
 *
 * @return A `MatcherResult` with `pass` and a message which can be handed to jest.
 */
function checkImages(
    snapshotImage: PngImage,
    receivedImage: PngImage,
    snapshotNumber: number,
    configuration: ToMatchImageSnapshotConfiguration,
): MatcherResult & { diffImage?: { data: Buffer, width: number, height: number } } {
    const {
        colorThreshold,
        detectAntialiasing,
        pixelThresholdAbsolute,
        pixelThresholdRelative,
    } = configuration;
    // Perform the actual image diff.
    const { pixels: changedPixels, image: diffImage } = diffImages({
        image1: receivedImage,
        image2: snapshotImage,
        colorThreshold,
        detectAntialiasing,
    });
    const expected = `stored snapshot ${snapshotNumber}`;
    const preamble = `${chalk.red("Received value")} does not match ${chalk.green(expected)}.`;
    if (typeof pixelThresholdAbsolute === "number" && changedPixels > pixelThresholdAbsolute) {
        return {
            pass: false,
            message: () =>
                `${preamble}\n\n` +
                `Expected less than ${chalk.green(`${pixelThresholdAbsolute} pixels`)} to have changed, ` +
                `but ${chalk.red(`${changedPixels} pixels`)} changed.`,
            diffImage,
        };
    }
    const snapshotImagePixels = snapshotImage.width * snapshotImage.height;
    const receivedImagePixels = receivedImage.width * receivedImage.height;
    const totalPixels = Math.max(snapshotImagePixels, receivedImagePixels);
    const changedRelative = changedPixels / totalPixels;
    if (typeof pixelThresholdRelative === "number" && changedRelative > pixelThresholdRelative) {
        const percentThreshold = (pixelThresholdRelative * 100).toFixed(2);
        const percentChanged = (changedRelative * 100).toFixed(2);
        return {
            pass: false,
            message: () =>
                `${preamble}\n\n` +
                `Expected less than ${chalk.green(`${percentThreshold}%`)} of the pixels to have changed, ` +
                `but ${chalk.red(`${percentChanged}%`)} of the pixels changed.`,
            diffImage,
        };
    }
    return { pass: true };
}

/**
 * A matcher for jest with compares a PNG image to a stored snapshot. Behaves similar to `.toMatchSnapshot()`.
 *
 * @param received The buffer from the call to `expect(...)`.
 * @param configuration The configuration object provided when initializing this library
 *     with a call to `jestScreenshot`.
 *
 * @return A `MatcherResult` usable by jest.
 */
export function toMatchImageSnapshot(
    received: Buffer,
    configuration: ToMatchImageSnapshotConfiguration,
): MatcherResult {
    // Check whether `this` is really the expected Jest configuration.
    if (!isJestTestConfiguration(this)) {
        throw new Error("Jest: Attempted to call `.toMatchImageSnapshot()` outside of Jest context.");
    }
    const { testPath, currentTestName, isNot } = this;
    if (isNot) {
        throw new Error("Jest: `.not` cannot be used with `.toMatchImageSnapshot()`.");
    }
    let { snapshotState } = this;
    const { _updateSnapshot } = snapshotState;
    const snapshotPath = getSnapshotPath(testPath, currentTestName, snapshotState, configuration);
    const reportPath = getReportPath(testPath, currentTestName, snapshotState, configuration);
    // Create the path to store the snapshots in.
    mkdirp(dirname(snapshotPath));
    // The image did not yet exist.
    if (!existsSync(snapshotPath)) {
        // If the user specified `-u`, or was running in interactive mode, write the new
        // snapshot to disk and let the test pass.
        if (_updateSnapshot === "new" || _updateSnapshot === "all") {
            snapshotState.added++;
            writeFileSync(snapshotPath, received);
            return { pass: true };
        }
        // Otherwise fail due to missing snapshot.
        return {
            pass: false,
            message: () => `New snapshot was ${chalk.red("not written")}. ` +
                `The update flag must be explicitly passed to write a new snapshot.\n\n` +
                `This is likely because this test is run in a continuous integration (CI) environment ` +
                `in which snapshots are not written by default.`,
        };
    }
    // Decode the new image and read the snapshot.
    const snapshotImage = readPngFileSync(snapshotPath);
    const receivedImage = decode(received);
    // Perform the actual diff of the images.
    const snapshotNumber = snapshotState._counters.get(currentTestName) || 1;
    const { pass, message, diffImage } = checkImages(snapshotImage, receivedImage, snapshotNumber, configuration);
    // If the user specified `-u` and the snapshot changed, update the stored snapshot.
    if (!pass) {
        if (_updateSnapshot === "all") {
            snapshotState.updated++;
            writeFileSync(snapshotPath, received);
            return { pass: true };
        }
        if (reportPath) {
            mkdirp(reportPath);
            writeFileSync(join(reportPath, "received.png"), received);
            writeFileSync(join(reportPath, "snapshot.png"), readFileSync(snapshotPath));
            writePngFileSync(join(reportPath, "diff.png"), diffImage.data, diffImage);
        }
    }
    return { pass, message };
}
