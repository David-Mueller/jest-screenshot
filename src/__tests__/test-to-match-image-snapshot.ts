import { jestScreenshot } from "..";
import { toMatchImageSnapshot } from "../to-match-image-snapshot";
import { SnapshotState, JestTestConfiguration } from "../jest";
import { readFileSync, unlinkSync, existsSync } from "fs";

function getJestTestConfiguration(): JestTestConfiguration {
    let testConfiguration: JestTestConfiguration;
    expect.extend({
        storeSnapshotState() {
            testConfiguration = this as any;
            return {
                pass: true,
                message() { return ""; },
            };
        },
    });
    (expect as any)().storeSnapshotState();
    return testConfiguration;
}

describe("toMatchImageSnapshot", () => {
    beforeAll(() => {
        expect.extend(jestScreenshot());
    });

    describe("with no threshold provided", () => {
        it("detects a matching snapshot as matching", () => {
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-gradient.png`)).toMatchImageSnapshot();
            }).not.toThrowError();
        });

        [
            { colorThreshold: 0.0 },
            { colorThreshold: 0.1 },
            { colorThreshold: 0.5 },
            { colorThreshold: 0.6 },
        ].forEach(({ colorThreshold }) => {
            it(`with a color threshold of ${colorThreshold} detects the snapshot as not matching`, () => {
                expect.extend(jestScreenshot({ colorThreshold }));
                expect(() => {
                    expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
                }).toThrowErrorMatchingSnapshot();
            });
        });

        [
            { colorThreshold: 0.7 },
            { colorThreshold: 1.0 },
        ].forEach(({ colorThreshold }) => {
            it(`with a color threshold of ${colorThreshold} detects the snapshot as matching`, () => {
                expect.extend(jestScreenshot({ colorThreshold }));
                expect(() => {
                    expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
                }).not.toThrowError();
            });
        });
    });

    describe("with an absolute threshold provided", () => {
        const colorThreshold = 0.6;

        it("fails with an absolute threshold of 204", () => {
            const pixelThresholdAbsolute = 204;
            expect.extend(jestScreenshot({ colorThreshold, pixelThresholdAbsolute }));
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
        });

        it("passes with an absolute threshold of 205", () => {
            const pixelThresholdAbsolute = 205;
            expect.extend(jestScreenshot({ colorThreshold, pixelThresholdAbsolute }));
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).not.toThrowError();
        });
    });

    describe("with `.not` specified", () => {
        it("fails always", () => {
            expect(() => {
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).not.toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
        });
    });

    describe("when called outside of a jest unit test context", () => {
        it("fails", () => {
            expect(function() {
                toMatchImageSnapshot(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`), {});
            }.bind({})).toThrowErrorMatchingSnapshot();
        });
    });

    describe("with the snapshot not existing", () => {
        let testConfig: JestTestConfiguration;
        let originalUpdateSnapshot: "new" | "all" | "none";
        const snapshotToCreatePath = `${__dirname}/__snapshots__/test-to-match-image-snapshot-ts-to-match-image-snapshot-with-the-snapshot-not-existing-creates-the-snapshot-when-updating-is-enabled-1.snap.png`; // tslint:disable-line

        afterEach(() => {
            try {
                unlinkSync(snapshotToCreatePath);
            } catch (err) {
                return;
            }
        });

        beforeEach(() => {
            testConfig = getJestTestConfiguration();
            originalUpdateSnapshot = testConfig.snapshotState._updateSnapshot;
        });

        afterEach(() => {
            testConfig.snapshotState._updateSnapshot = originalUpdateSnapshot;
        });

        const test = it("fails when updating is disabled", () => {
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "none";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).toThrowErrorMatchingSnapshot();
        });

        it("creates the snapshot when updating is enabled", () => {
            expect(existsSync(snapshotToCreatePath)).toBe(false);
            expect(() => {
                testConfig.snapshotState._updateSnapshot = "all";
                expect(readFileSync(`${__dirname}/fixtures/red-rectangle-example-red.png`)).toMatchImageSnapshot();
            }).not.toThrowError();
            expect(existsSync(snapshotToCreatePath)).toBe(true);
        });
    });
});
