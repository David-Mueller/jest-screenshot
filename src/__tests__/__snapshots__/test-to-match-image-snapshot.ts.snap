// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchImageSnapshot when called outside of a jest unit test context fails 1`] = `"Jest: Attempted to call \`.toMatchImageSnapshot()\` outside of Jest context."`;

exports[`toMatchImageSnapshot with \`.not\` specified fails always 1`] = `"Jest: \`.not\` cannot be used with \`.toMatchImageSnapshot()\`."`;

exports[`toMatchImageSnapshot with an absolute threshold provided fails with an absolute threshold of 204 1`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m204 pixels[39m to have changed, but [31m205 pixels[39m changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0 detects the snapshot as not matching 1`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m53.24%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.1 detects the snapshot as not matching 1`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m45.44%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.5 detects the snapshot as not matching 1`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m12.52%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with no threshold provided with a color threshold of 0.6 detects the snapshot as not matching 1`] = `
"[31mReceived value[39m does not match [32mstored snapshot 1[39m.

Expected less than [32m0.00%[39m of the pixels to have changed, but [31m4.10%[39m of the pixels changed."
`;

exports[`toMatchImageSnapshot with the snapshot not existing fails when updating is disabled 1`] = `
"New snapshot was [31mnot written[39m. The update flag must be explicitly passed to write a new snapshot.

This is likely because this test is run in a continuous integration (CI) environment in which snapshots are not written by default."
`;
